chapter_15_advanced_forms.asciidoc

```diff
@@ -508,10 +317,48 @@ it as an application-layer constraint:
 $ pass:quotes[*python manage.py makemigrations*]
 Migrations for 'lists':
   src/lists/migrations/0005_alter_item_options_alter_item_unique_together.py
-    - Change Meta options on item
     - Alter unique_together for item (1 constraint(s))
 ----
 
+Now let's run the migration:
+
+[subs="specialcharacters,macros"]
+----
+$ pass:quotes[*python manage.py migrate*]
+----
+
+When you run the migration, you may encounter the following error:
+[subs="specialcharacters,macros"]
+----
+$ pass:quotes[*python manage.py migrate*]
+Operations to perform:
+  Apply all migrations: auth, contenttypes, lists, sessions
+Running migrations:
+  Applying lists.0005_alter_item_unique_together...Traceback (most recent call last):
+[...]
+sqlite3.IntegrityError: UNIQUE constraint failed: lists_item.list_id, lists_item.text
+
+[...]
+django.db.utils.IntegrityError: UNIQUE constraint failed: lists_item.list_id, lists_item.text
+----
+
+The problem is that we have at least one database record which used to be valid
+but after introducing our new constraint, the `unique_together`, it's no longer
+compatible.
+
+Just delete `src/db.sqlite3` and run the migration again. We can do this,
+because it's only a dev environment, and the data saved in the database isn't
+important.
+
+However, if we already have records we want to keep (which would be the case in
+a production environment, where we need to keep our users' data) we would need
+to do a bit more work and come up with a migration strategy for dealing with
+invalid--duplicate in this case--records.
+// CSANAD: I wanted to mention some articles from credible sources explaining
+//         migration strategies, modifying migration scripts, etc
+//         but all I found were either outdated, or clearly chatgpt-generated
+// articles.
```

Comment:
hmmm i think we do mention prod migrations at some point... let me look it up.

Reply:
yes it's mentioned in chapter 17, when we try and deploy the latest code to prod, and there's an appendix on data migrations.  http://www.obeythetestinggoat.com/book/chapter_deploying_validation.html#_what_to_do_if_you_see_a_database_error

Reply:
```suggestion
In <chapter_deploying_validation>>, we'll deploy our new code to production,
and discuss what to do if we run into migrations and data integrity issues at that point.
```